algoliasearch/index.go

```diff
@@ -1,63 +1,71 @@
 package algoliasearch
 
 import (
-	"errors"
+	"fmt"
 	"net/url"
-	"reflect"
-	"strconv"
 	"strings"
 	"time"
 )
 
 // Index is the structure used to manipulate an Algolia index.
 type Index struct {
-	client      *Client
-	name        string
-	nameEncoded string
+	client *Client
+	name   string
+	route  string
 }
 
 // NewIndex instantiates a new Index. The `name` parameter corresponds to the
 // Algolia index's name while the `client` is used to connect to the Algolia
 // API.
 func NewIndex(name string, client *Client) *Index {
 	return &Index{
-		client:      client,
-		name:        name,
-		nameEncoded: client.transport.urlEncode(name),
+		client: client,
+		name:   name,
+		route:  "/1/indexes/" + client.transport.urlEncode(name),
 	}
 }
 
 // Delete deletes the Algolia index.
-func (i *Index) Delete() (interface{}, error) {
-	return i.client.transport.request("DELETE", "/1/indexes/"+i.nameEncoded, nil, write)
+func (i *Index) Delete() (res DeleteTaskRes, err error) {
+	path := i.route
+	err = i.client.request(&res, "DELETE", path, nil, write)
+	return
 }
 
 // Clear removes every record from the Algolia index.
-func (i *Index) Clear() (interface{}, error) {
-	return i.client.transport.request("POST", "/1/indexes/"+i.nameEncoded+"/clear", nil, write)
+func (i *Index) Clear() (res UpdateTaskRes, err error) {
+	path := i.route + "/clear"
+	err = i.client.request(&res, "POST", path, nil, write)
+	return
 }
 
 // GetObject retrieves the object as an interface representing the JSON-encoded
 // object. The `objectID` is used to uniquely identify the object in the index
 // while the `attribute` (optional) is a string containing comma-separated
 // attributes that you want to retrieve. If this parameter is omitted, all the
 // attributes are returned.
-func (i *Index) GetObject(objectID string, attribute ...string) (interface{}, error) {
-	v := url.Values{}
-	if len(attribute) > 1 {
-		return nil, errors.New("Too many parameters")
-	} else if len(attribute) > 0 {
-		v.Add("attribute", attribute[0])
+func (i *Index) GetObject(objectID string, attributes []string) (object Object, err error) {
+	// URL-encode the objectID and the attributes
+	objectID = i.client.transport.urlEncode(objectID)
+	var values url.Values
+	for _, attr := range attributes {
+		values.Add("attributes", attr)
```

Comment:
I moved `encodeParams` to `utils.go` and use it wherever I can now.
I've also deleted the `urlEncode` function from `transport.go` as it was only used by the `Index` and it was a simple call to `url.QueryEscape`. No more "i.client.transport"-like calls.